### ISSUE TITLE
libevent & openssl: After SSL connection is disconnected, libevent cannot reconnect.

### ISSUE THREAD

**Issue by tech-a-go-go** [libevent/libevent#975]
Hi gurus,

I'm getting TLS handshake error after ssl connection is disconnected and connected again.
( I'm not sure this is caused by libevent or openssl. I'm sorry if this is an openssl issue or my code is just bad. I don't know where else I can ask 😢  )
 
Here is my super simple https client. What it does is just 2 things: 

1. access https://127.0.0.1 every second
2. reconnect when the connection is disconnected
```c++
#include <event2/bufferevent_ssl.h>
#include <evhttp.h>
#include <openssl/err.h>
#include <openssl/rand.h>
#include <openssl/ssl.h>
#include <unistd.h>
#include <iostream>

using namespace std;

SSL_CTX *sslCtx;
SSL *ssl;
event_base *eventBase;
evhttp_request *request;
bufferevent *bev;
evhttp_connection *conn;

void on_request_completed(struct evhttp_request *req, void *obj) {
    if (!req) printf("Disconnected !\n");
    event_base_loopbreak(eventBase);
}

void on_connection_closed(struct evhttp_connection * /*conn*/, void *obj) {
    if (conn) conn = nullptr;
    if (bev) bev = nullptr;
    if (ssl) ssl = nullptr;
}

void init() {
    SSL_library_init();
    ERR_load_crypto_strings();
    SSL_load_error_strings();
    OpenSSL_add_all_algorithms();
    sslCtx = SSL_CTX_new(SSLv23_method());
    eventBase = event_base_new();
}

void access() {
    request = evhttp_request_new(on_request_completed, nullptr);
    evhttp_add_header(request->output_headers, "Host", "127.0.0.1:443");
    if (conn) {
        event_base_loop(eventBase, EVLOOP_NONBLOCK);
    }

     // make new connection only when first access and connection is disconnected.
    if (conn == nullptr) {
        ssl = SSL_new(sslCtx);
        bev = bufferevent_openssl_socket_new(eventBase, -1, ssl,
                                             BUFFEREVENT_SSL_CONNECTING,
                                             BEV_OPT_CLOSE_ON_FREE | BEV_OPT_DEFER_CALLBACKS);
        bufferevent_openssl_set_allow_dirty_shutdown(bev, 1);

        conn = evhttp_connection_base_bufferevent_new(eventBase, nullptr, bev,
                                                      "127.0.0.1", 443);
        evhttp_connection_set_closecb(conn, on_connection_closed, nullptr);
    }
    evhttp_make_request(conn, request, EVHTTP_REQ_GET, "https://127.0.0.1:443/");
    event_base_dispatch(eventBase);
}

int main() {
    init();
    while (1) {
        printf("access\n");
        access();
        sleep(1);
    }
    return 0;
}
```
* SSL_CTX and event_base variables are created only once when app starts.
* SSL, evhttp_request, bufferevent and evhttp_connection variables are created every time when connection is disconnected (and first access).

It works very well until the connection is disconnected.

While this app is accessing the localhost server, I restarted the server to disconnect on purpose.
This app tries to reconnect to the server (so far so good) but the server throws "TLS handshake error" when the server is up.

I googled libevent docs, debug and did everything I can do but no luck... 

Please help me out. 

Cheers

**Comment by azat** [comment]
In a nut shell you code should work (even though it has some issues), and actually I think that the problem is that libevent and your code uses ABI incompatible openssl versions (i.e. 1.1 and 1.0), can you describe your enviroment? (libevent version, os, openssl version, openssl devel package version if any)

**Comment by azat** [comment]
And just a my assumption, looks like your code assumes openssl 1.0, I'm not 100% sure, but the code below is required only in 1.0 version, 1.1 doing it by itself (and actually if you openssl compiled with `OPENSSL_NO_DEPRECATED` then you will not have such symbols at all)

```c
    SSL_library_init();
    ERR_load_crypto_strings();
    SSL_load_error_strings();
    OpenSSL_add_all_algorithms();
```

**Comment by tech-a-go-go** [comment]
Hi @azat ,

Thank you for your quick response !

My os is CentOS 7.7
```
$ cat /etc/redhat-release 
CentOS Linux release 7.7.1908 (Core)
$ uname -r
3.10.0-1062.12.1.el7.x86_64
```

My libevent version is 2.2. 
I checked out libevent from github last week, compiled **master branch** and installed.
```
$ ls
libevent-2.2.so.1      libevent_core-2.2.so.1      libevent_extra-2.2.so.1      libevent_openssl-2.2.so.1      libevent_pthreads-2.2.so.1      pkgconfig
libevent-2.2.so.1.0.0  libevent_core-2.2.so.1.0.0  libevent_extra-2.2.so.1.0.0  libevent_openssl-2.2.so.1.0.0  libevent_pthreads-2.2.so.1.0.0
...
```

As you said, my openssl version is 1.0.2.  ( opensslv.h says "OpenSSL 1.0.2k-fips  26 Jan 2017" )
```
$ yum list installed | grep openssl
openssl.x86_64                          1:1.0.2k-19.el7                @base    
openssl-devel.x86_64                    1:1.0.2k-19.el7                @base    
openssl-libs.x86_64                     1:1.0.2k-19.el7                @base    
```
I commented out these 4 lines you mentioned and started app, then it crashed so my openssl version is 1.0 for sure =)
```
    SSL_library_init();
    ERR_load_crypto_strings();
    SSL_load_error_strings();
    OpenSSL_add_all_algorithms();
```

Thanks a billion.

**Comment by tech-a-go-go** [comment]
hi @azat 

I've installed openssl 1.1 and tried everything again.
I have two news, one good and one bad...

The good one is "TLS handshake error". Thank you so much !!

The bad one is the **on_connection_closed** callback (set by evhttp_connection_set_closecb(conn, on_connection_closed, nullptr)) is not called even though the connection is disconnected.

When I restart the localhost server, now my libevent gets silently disconnected and reconnect again.

The connection closed callback doesn't work with openssl 1.1 ?? 

Thank you

**Comment by azat** [comment]
>The bad one is the on_connection_closed callback (set by evhttp_connection_set_closecb(conn, on_connection_closed, nullptr)) is not called at all even though the connection is disconnected.

It is called only when the connection had been established before, so seems that you tried it without server and saw the "problem"

**Comment by tech-a-go-go** [comment]
Hi @azat 

Oops, I'm sorry. Yes, evhttp_connection_set_closecb is working pretty well with openssl 1.1 !

The request arg of on_request_completed function below is no longer null even though the connection is disconnected.
```
void on_request_completed(struct evhttp_request *request, void *obj) {
    if (!request) printf("Disconnected !\n");
    event_base_loopbreak(eventBase);
}
```
Are there any APIs I can query so that I can tell the connection is disconnected ?
Or should I manage it by myself with the on_connection_closed function ?

The reason I'm asking about the connection stuff is that I'm making a http(s) client pool now.
I want to remove a client from the pool when the client's connection is disconnected. That's why the connection state is very important.

Cheers

**Comment by tech-a-go-go** [comment]
HI @azat 

I've decided to remove a http client from the pool when on_connection_closed is called.
The libevent instance of the http client is no longer needed so I called bufferevent_free(bufferevent) but ended up with the following Assertion error.
```
[err] bufferevent.c:713: Assertion bufev_private->refcnt > 0 failed in bufferevent_decref_and_unlock_
./run.sh: line 16: 126828 Aborted                 (core dumped) ./client
```

I don't have to free bufferevent after on_connection_closed  is called ?? 
(Has it already been freed automatically when the connection got disconnected ? )

Cheers

**Comment by azat** [comment]
>(Has it already been freed automatically when the connection got disconnected ? )

[Yes](http://libevent.org/doc/http_8h.html#ab6a2f82ff7323986705ead089a2d3f58)

> a bufferevent to use for connecting to the server; if NULL, a socket-based bufferevent will be created. This buffrevent will be freed when the connection closes. It must have no fd set on it.

**Comment by azat** [comment]
>The request arg of on_request_completed function below is no longer null even though the connection is disconnected so that "Disconnected !" log is not shown.

Version of openssl does not related to this, it should always work like this

>Are there any APIs I can query so that I can tell the connection is disconnected ?

No

>Or should I manage it with the on_connection_closed function by myself ?

Actually this will trigger the callback only if the connection has been established before, are you sure that this is what you need?

*P.S. Callbacks in http are somewhat tricky (sigh)*

**Comment by tech-a-go-go** [comment]
@azat 

I got everything I need.

Thank you so much !
